### PR TITLE
Add TP support with vLLM FusedMoE kernel

### PR DESCRIPTION
**Pull request by NicoGrande**
This PR adds support for TP using `tpu-inference` GMM_V2 kernel. 

To do this, we add logic to pad the MLP dimension of MoE kernels to the closest power of 2 that satisfies the divisibility constraint of the GMM_V2 kernel. Additionally, we add logic to `model_creation_utils.py` to correctly handle the padding case when loading a model checkpoint for inference. 

# Tests

E2E runs with `vllm_decode.py`. E2E GRPO runs. New unit tests to validate the logic in `model_creation_utils.py`.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by hengtaoguo — APPROVED** [review]

**Review by khatwanimohit — APPROVED** [review]
LGTM

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py [hunk 1]
```diff
@@ -91,6 +97,19 @@ def generate_maxtext_config(vllm_config: VllmConfig, mesh: Mesh) -> pyconfig.Hyp
     )
     overrides["base_num_kv_heads"] = max_tp_size
 
+  # Pad the hidden size of MoE models if the MLP dimension is less than expected by the GMM_v2 kernel in tpu-inference
+  hidden_size = getattr(vllm_config.model_config.hf_config, "moe_intermediate_size", None)
+  padded_hidden_size = hidden_size
+  num_lanes = pltpu.get_tpu_info().num_lanes
+  total_tp_size = jnp.prod([mesh.shape[axis_name] for axis_name in ShardingAxisName.MLP_TENSOR]).item()
+
+  if hidden_size is not None and tp_axis_name == "model" and (hidden_size // total_tp_size) % (2 * num_lanes) != 0:
```
Comment (NuojCheng) [inline src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: where the 2 * num_lanes comes from? maybe comment

File: src/maxtext/utils/model_creation_utils.py [hunk 2]
```diff
@@ -51,18 +50,99 @@ def _is_orbax_array_metadata(x):
...
+            f"shards_along_axis={shards_along_axis} (sharding spec entry {partition!r})."
+        )
+      per_shard_extra = extra // shards_along_axis
+      pad_width = [(0, 0)] * arr.ndim
+      pad_width[axis] = (0, per_shard_extra)
+
+      def _pad_local(x):
+        return jnp.pad(x, pad_width)
+
+      return jax.shard_map(_pad_local, mesh=sharding.mesh, in_specs=spec, out_specs=spec, check_vma=False)(arr)
```
Comment (NuojCheng) [inline src/maxtext/utils/model_creation_utils.py:117]: why use shard_map in this case? I don't see collectives happening inside `_pad_local` function...

File: src/maxtext/utils/model_creation_utils.py [hunk 3]
```diff
@@ -51,18 +50,99 @@ def _is_orbax_array_metadata(x):
...
+
+  pad_width = [(0, 0)] * arr.ndim
```
Comment (NuojCheng) [inline src/maxtext/utils/model_creation_utils.py]: nit: move `pad_width = [(0, 0)] * arr.ndim` above line 102?

File: src/maxtext/utils/model_creation_utils.py [hunk 4]
```diff
@@ -74,27 +154,163 @@ def _expand_checkpoint_to_model_shapes(ckpt_arr, model_arr):
...
+  slice of both wi_0 and wi_1. It also applies any necessary MLP-dim padding
+  on a per-shard basis to satisfy kernel constraints.
+  """
+  if not hasattr(ckpt_tree, "items") or not hasattr(model_arrays_tree, "items"):
+    return ckpt_tree
+  new_ckpt = {}
+  for k, v in ckpt_tree.items():
+    if k in ("wi_0", "wi_1") and "wi" in model_arrays_tree:
+      continue
+    new_ckpt[k] = _fuse_moe_weights(v, model_arrays_tree.get(k, {}))
```
Comment (NuojCheng) [inline src/maxtext/utils/model_creation_utils.py]: nit: I might use `jax.tree.map` instead of recursion if possible

File: src/maxtext/utils/model_creation_utils.py [hunk 5]
```diff
@@ -634,6 +863,23 @@ def _adjust_target_for_moe_fusion(target, meta_tree, is_nnx):
...
 
+        # Free memory used by initial sharded_state before restore, to make room for the incoming checkpoint arrays.
+        def _free_device_memory(node):
+          if isinstance(node, nnx.Variable):
+            # Do NOT delete RNG state; it isn't in the checkpoint and won't be replaced
+            if isinstance(node, nnx.RngState):
+              return node
+            val = node.value
+          else:
+            val = node
```
Comment (NuojCheng) [inline src/maxtext/utils/model_creation_utils.py:897]:
```
if isinstance(node, nnx.Variable) and not isinstance(node, nnx.RngState):
  val = node.value
else:
val = node
```

File: tests/unit/model_creation_utils_test.py [hunk 6]
```diff
@@ -115,15 +133,35 @@ def test_scanned_ckpt_unscanned_model_raises_error(self):
...
+  def test_same_rank_shape_mismatch_indivisible_falls_back_to_replicated(self):
+    """Stored dim not divisible by mesh axis size: fall back to fully-replicated."""
+    if jax.device_count() < 2:
+      self.skipTest("Requires >=2 devices to construct an indivisible mesh axis.")
+    devices = jax.local_devices()[:2]
+    multi_mesh = jax.sharding.Mesh(devices, ("x",))
+    sharding = jax.sharding.NamedSharding(multi_mesh, jax.sharding.PartitionSpec(None, "x", None))
+    # stored axis 1 = 3, mesh "x" = 2 -> 3 % 2 != 0 -> can't shard, must replicate.
+    stored_shape = (4096, 3, 128)
+    model_shape = (4096, 4, 128)
```
Comment (NuojCheng) [inline tests/unit/model_creation_utils_test.py]: Can we make these sizes smaller?

File: tests/unit/model_creation_utils_test.py [hunk 7]
```diff
@@ -132,13 +170,302 @@ def test_same_shape_no_modification(self):
...
+    """Scanned ckpt + scanned model + KV padding, divisible stored shape: keep sharding."""
     stored_shape = (94, 4096, 4, 128)
     model_shape = (94, 4096, 64, 128)
     fixed = self._run_fix(stored_shape, model_shape)
     arg = fixed["kernel"]
-    self.assertIsNone(arg.global_shape)
+    self.assertEqual(arg.global_shape, stored_shape)
+
+
+class TestAlignCheckpointToModelShapes(unittest.TestCase):
```
Comment (NuojCheng) [inline tests/unit/model_creation_utils_test.py:185]: can we make them tpu_only tests? (or maybe they are cpu tests...)

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py (hunk 1 above)
Comment (NicoGrande) [reply src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: Added a comment pointing to the validation performed in the GMM_v2 kernel

File: src/maxtext/utils/model_creation_utils.py (hunk 2 above)
Comment (NicoGrande) [reply src/maxtext/utils/model_creation_utils.py:117]: We need to apply the padding in a per-shard manner. If my model dimension is 768 and I am using TP=4, I would expect each shard of dimension 192 to be padded to 256. The shard map allows me to do this without having to aggregate the global tensor in one device which can lead to OOMs. I tried removing the shardmap and saw gibberish outputs.

File: src/maxtext/utils/model_creation_utils.py (hunk 3 above)
Comment (NicoGrande) [reply src/maxtext/utils/model_creation_utils.py]: Done

File: src/maxtext/utils/model_creation_utils.py (hunk 4 above)
Comment (NicoGrande) [reply src/maxtext/utils/model_creation_utils.py]: Done

File: src/maxtext/utils/model_creation_utils.py (hunk 5 above)
Comment (NicoGrande) [reply src/maxtext/utils/model_creation_utils.py:897]: Done

File: tests/unit/model_creation_utils_test.py (hunk 6 above)
Comment (NicoGrande) [reply tests/unit/model_creation_utils_test.py]: Yes, reduced to 256

File: tests/unit/model_creation_utils_test.py (hunk 7 above)
Comment (NicoGrande) [reply tests/unit/model_creation_utils_test.py:185]: Added tpu_only tags :)

File: src/maxtext/utils/model_creation_utils.py (hunk 2 above)
Comment (NuojCheng) [reply src/maxtext/utils/model_creation_utils.py:117]: very cool

**Review by NuojCheng — APPROVED** [review]
